Site moderator promotions

tessalt opened the pull request:
<!--

Thank you for submitting a pull request! Please note that by contributing to
Coral, you agree to our Code of Conduct: http://code-of-conduct.voxmedia.com/

Before submitting your Pull Request (or PR), please verify that:

* [ ] Your code is up-to-date with the base branch
* [ ] You've successfully run `npm run test` locally

Refer to CONTRIBUTING.MD for more details.

  https://github.com/coralproject/talk/blob/master/CONTRIBUTING.md

-->

## What does this PR do?

Implements changes needed to allow site moderators limited role change permissions, allowing moderator's to update a user's role in the following cases:
- moderators can only change role *to* moderator
- scoped moderators can only create scoped moderators
- scoped moderators can only add moderators to sites they already moderate
- moderators cannot demote admins

The complexity of these conditions may suggest another solution is required, such as the creation of "site admins". PR as a discussion point more than proposed feature. 

<!--

In this section, you should be describing what other Github issues or tickets
that this PR is designed to addressed.

Any related Github issue should be linked by adding its URL to this section.

-->

## What changes to the GraphQL/Database Schema does this PR introduce?

<!--

In this section, you should describe any changes to be made to the GraphQL
schema file (located https://github.com/coralproject/talk/blob/master/src/core/server/graph/schema/schema.graphql) or any
database model (located as types in the https://github.com/coralproject/talk/blob/master/src/core/server/models directory).

If no changes were added to the GraphQL/Database Schema as a part of this PR,
simply write "None".

-->

## How do I test this PR?

<!--

In this section, you should be describing any manual testing that can be used to
verify features introduced or bugs fixed in this PR.

 -->

tessalt committed:
allow site moderators to promote other site moderators

tessalt committed:
add todo

tessalt committed:
allow modifying roles and scopes in one mutation

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7716,7 +7722,7 @@ type Mutation {
   """
   updateUserRole(input: UpdateUserRoleInput!): UpdateUserRolePayload!
-    @auth(roles: [ADMIN])
+    @auth(roles: [ADMIN, MODERATOR])
```
We should not re-use the `updateUserRole` for this operation.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7725,7 +7731,7 @@ type Mutation {
   updateUserModerationScopes(
     input: UpdateUserModerationScopesInput!
-  ): UpdateUserModerationScopesPayload! @auth(roles: [ADMIN])
+  ): UpdateUserModerationScopesPayload! @auth(roles: [ADMIN, MODERATOR])
```
We should not re-use the `updateUserModerationScopes` for this operation.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -655,21 +656,50 @@ export async function updateUsernameByID(
+  if (user.role === GQLUSER_ROLE.MODERATOR) {
+    if (role !== GQLUSER_ROLE.MODERATOR) {
+      throw new Error("Moderators can only promote to moderator");
+    }
```
Authorization must generally not be completed within the service functions.

nick-funk committed:
Complete the existing site moderator role management

- Site moderators can promote new site moderators
- Org mods can promote/demote commenters, staff, site mods
- Admins can do anything they want

CORL-1364

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7772,7 +7811,13 @@ type Mutation {
+  """
+  promoteUser will allow you to promote a user to a moderator.
+  """
+  promoteUser(input: PromoteUserInput!): PromoteUserPayload!
```
```suggestion
  promoteUserRole(input: PromoteUserRoleInput!): PromoteUserRolePayload!
```

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
+export async function promoteUser(
+  mongo: Db,
+  tenant: Tenant,
+  user: Pick<User, "id" | "role" | "moderationScopes">,
```
I'd suggest calling this `viewer` to disassociate this from the user we're promoting a bit further.
```suggestion
  viewer: Pick<User, "id" | "role" | "moderationScopes">,
```

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
+  // you can only promote users if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(user.role)) {
+    throw new Error("you have insufficient permissions to promote users");
+  }
```
This is not required as it's already handled by the `@auth` directive in the schema.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
+
+  // you can't promote a user that is already an admin
+  const userToPromote = await retrieveUser(mongo, tenant.id, userIDToPromote);
+  if (userToPromote?.role === GQLUSER_ROLE.ADMIN) {
```
This seems in error. If you need to use `?` in `userToPromote?.role`, it implies you should check for `if (!userToPromote) { throw new UserNotFound }` instead.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -6738,6 +6738,45 @@ type UpdateUserRolePayload {
+  """
+  role is the `USER_ROLE` that the User should be set to.
+  """
+  role: USER_ROLE!
```
This promotion edge is already exclusively used to promote users to moderator roles. As such it shouldn't accept a role and should assume moderator instead.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -6738,6 +6738,45 @@ type UpdateUserRolePayload {
+  moderationScopes defines the scopes that a given moderator should be
+  limited to
+  """
+  moderationScopes: ModerationScopesInput
```
These should also be implied, as the user should only be able to promote other users to the same level as them.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
+  }
+
+  return await updateUserRole(mongo, tenant.id, userID, role);
+}
```
These changes should be reverted because we should only be affecting the `promoteUser` instead.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +692,133 @@ export async function updateUsernameByID(
+      userIDToPromote,
+      moderationScopes
+    );
+  }
```
One DB call instead of two please!

nick-funk committed:
Make it so user's can't reassign their own roles

CORL-1364

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -682,21 +772,38 @@ export async function updateModerationScopes(
+  // you can only change scopes if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(viewer.role)) {
+    throw new Error("you have insufficient permissions to promote users");
+  }
```
This check is completed by the schema level and should not be duplicated here.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -682,21 +772,38 @@ export async function updateModerationScopes(
+    viewer.moderationScopes.siteIDs?.length !== 0
+  ) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
```
This check should be completed at the schema level and should not be duplicated here.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +659,110 @@ export async function updateUsernameByID(
+    !_.isEqual(
+      _.sortBy(viewer.moderationScopes?.siteIDs),
+      _.sortBy(scopesResult.moderationScopes?.siteIDs)
+    )
```
You should not use the default import here for lodash and instead import the named functions.

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7825,7 +7853,13 @@ type Mutation {
   """
   updateUserRole(input: UpdateUserRoleInput!): UpdateUserRolePayload!
-    @auth(roles: [ADMIN])
+    @auth(roles: [ADMIN, MODERATOR])
```
Add `unscoped` to require moderators to be unscoped when using this mutation:

```suggestion
    @auth(roles: [ADMIN, MODERATOR], unscoped: true)
```

wyattjoh commented on src/core/server/graph/schema/schema.graphql:
```diff
@@ -7834,7 +7868,7 @@ type Mutation {
   updateUserModerationScopes(
     input: UpdateUserModerationScopesInput!
-  ): UpdateUserModerationScopesPayload! @auth(roles: [ADMIN])
+  ): UpdateUserModerationScopesPayload! @auth(roles: [ADMIN, MODERATOR])
```
Add `unscoped` to require moderators to be unscoped when using this mutation:

```suggestion
  ): UpdateUserModerationScopesPayload! @auth(roles: [ADMIN, MODERATOR], unscoped: true)
```

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +659,110 @@ export async function updateUsernameByID(
+  // you can only update users if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(viewer.role)) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
```
Not required and should be removed as it's currently duplicating logic from the schema.

wyattjoh commented on src/core/server/services/users/users.ts:
```diff
@@ -658,21 +659,110 @@ export async function updateUsernameByID(
+    viewer.moderationScopes.siteIDs?.length !== 0
+  ) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
```
This should be removed when `unscoped: true` is added to the `@auth` directive instead.

wyattjoh commented on src/core/client/admin/components/UserRole/UserRoleChangeContainer.tsx:
```diff
@@ -80,22 +137,52 @@ const UserRoleChangeContainer: FunctionComponent<Props> = ({
+    default:
+      return GQLUSER_ROLE.COMMENTER;
+  }
+}
```
Using the `GQLUSER_ROLE_RL` is important here, and should not be converted to a `GQLUSER_ROLE`. When referring to the _value_, you should always use `GQLUSER_ROLE` (Like when you're saying `viewer.role !== GQLUSER_ROLE.ADMIN`). When _typeing_ it, you should always use `GQLUSER_ROLE_RL`, as it ensures the type system will be forwards compatible with new role types.

wyattjoh commented on src/core/client/admin/components/UserRole/UserRoleChangeContainer.tsx:
```diff
@@ -80,22 +137,52 @@ const UserRoleChangeContainer: FunctionComponent<Props> = ({
+        viewer.moderationScopes
+          ? viewer.moderationScopes.sites?.map((s) => s.id)
+          : null
+      }
```
It feels like there's a lot of leakage and duplication of logic in this container component and it's associated control component. Ideally, we should be passing more explicitly what modes to enable for the `UserRoleChange` component rather than giving it all the context to make it's own decisions.

nick-funk committed:
User GQLUSER_ROLE_RL generated types for roles

CORL-1364

nick-funk committed:
Clean up guard clauses and auth directives

CORL-1364

nick-funk committed:
Simplify viewer dependencies in UserRoleChange

CORL-1364

wyattjoh committed:
Merge branch 'master' into spike/site-moderator-promotions

wyattjoh commented on src/core/client/admin/components/UserRole/PromoteUserRoleMutation.ts:
```diff
@@ -0,0 +1,45 @@
+let clientMutationId = 0;
+
+const PromoteUserRoleMutation = createMutation(
+  "updateUserModerationScopes",
```
Update to `promoteUserRole`

wyattjoh commented on src/core/client/admin/components/UserRole/SiteModeratorModal.tsx:
```diff
@@ -90,7 +94,11 @@ const SiteModeratorModal: FunctionComponent<Props> = ({
+                  <SiteModeratorModalSiteFieldContainer
+                    query={query}
+                    viewerScoped={viewerScoped}
+                    viewerSites={viewerSites}
```
Query from `query` instead

wyattjoh commented on src/core/client/admin/components/UserRole/UserRoleChange.tsx:
```diff
@@ -22,53 +25,196 @@ import styles from "./UserRoleChange.css";
+export function canChangeRole(
+  viewerRole: RoleDescription,
+  existingRole: RoleDescription,
+  newRole: RoleDescription
```
```suggestion
  viewer: RoleDescription,
  user: RoleDescription,
  target: RoleDescription
```

wyattjoh commented on src/core/client/admin/components/UserRole/UserRoleChange.tsx:
```diff
@@ -22,53 +25,196 @@ import styles from "./UserRoleChange.css";
+    }
+
+    return viewer.moderationScopes.sites ? viewer.moderationScopes.sites : [];
+  }, [viewer]);
```
```suggestion
  const viewerSites = viewer.moderationScopes?.sites || [];
```

Commit:
fix: changes to promote/demote flow

Commit:
fix: revert changes to perm

Commit:
fix: revert changes to updateRole

Commit:
fix: revert changes to updateUserModerationScopes

kodiakhq[bot] committed:
Merge branch 'master' into spike/site-moderator-promotions
